### DISCUSSION TITLE
Does the "mon-host" configuration need to be exactly the same as the one in ceph.conf?

### DISCUSSION THREAD

**Discussion by dennis9352** [ceph/go-ceph#1049] [Q&A]
I tried connecting to a remote Ceph-mon server that is not in the same network, but it did not proceed from `rados.conn.Connect()`. Then, I tried running it from a server in the same network, and it worked.

If I need to use a proxy in the middle, the mon-host will change, so is there a way to separately configure the address to connect to ceph-mon and the mon-host setting in `ceph.conf`? To illustrate the situation I tested:

`192.168.0.60 (my computer) -> 192.168.0.240:3300 -> 172.168.0.230:3300 (ceph)`

In this case, I would have to set the `mon-host` to `192.168.0.240:3300`, which wouldn't match the `ceph.conf`.

However, if I set it to `172.168.0.231 -> 172.168.0.230:3300 (ceph)`, it seems to match the `ceph.conf` and the connection works. Is there an option to separate these configurations? Or is there a configuration mistake I might have made?

```
func StartCeph(address string, port int, cephKey string) error {
	if err := TestConnection(address, port); err != nil {
		return fmt.Errorf("failed to connect to ceph %s:%d: %w", address, port, err)
	}

	Instance = &Ceph{mu: &sync.Mutex{}}

	conn, err := rados.NewConn()
	if err != nil {
		return fmt.Errorf("failed to create new RADOS connection: %w", err)
	}

	Instance.conn = conn

	args := []string{
		"--mon-host", fmt.Sprintf("%s:%d", address, port),
		"--key", cephKey,
		"--client-mount-timeout", "30",
		"--debug-ms", "20",
	}

	if err := conn.ParseCmdLineArgs(args); err != nil {
		return fmt.Errorf("failed to parse configuration options: %w", err)
	}

	if err := conn.Connect(); err != nil {
		return fmt.Errorf("failed to connect to Ceph cluster: %w", err)
	}

	return nil
}
```

**Comment by SQUADRON-LEADER** [comment]
Hi, @dennis9352  `mon_host` is not just the initial bootstrap address. After the client connects to the monitor, Ceph returns the monitor map containing the real monitor addresses.
So if your client is outside the network and the monitor map contains internal IPs like `172.x.x.x`, the client will later try to connect to those internal addresses and `conn.Connect()` may hang/fail.

In your proxy case:

`192.168.0.60 -> 192.168.0.240:3300 -> 172.168.0.230:3300`

even if bootstrap works through `192.168.0.240`, the monitor map still advertises `172.168.0.230`, which your client cannot reach directly.

`--mon-host` and the monitor addresses advertised by Ceph cannot really be separated cleanly from the client side. The proper fix is usually one of these:

1. Configure Ceph monitors to advertise reachable public addresses (`public_addr` / public network).
2. Use NAT/port-forwarding so the advertised monitor IPs are reachable externally.
3. Use a VPN between the client and Ceph network.
4. Put a TCP proxy/load balancer in front of all monitors and make the monitor map advertise those proxy addresses.

Your Go code itself looks mostly fine — the issue is more about the monitor map/network routing than `rados.NewConn()` or `ParseCmdLineArgs()`.

You can confirm this by enabling more debug logs:

```go
"--debug-ms", "20",
"--debug-monc", "20",
```

and checking whether the client first connects successfully and then switches to unreachable monitor IPs from the monmap.
